nada_mir/proto/nillion/nada/v1/mir.proto

```diff
@@ -0,0 +1,113 @@
+syntax = "proto3";
+
+package nillion.nada.mir.v1;
+
+import "nillion/nada/v1/operations.proto";
+import "nillion/nada/v1/types.proto";
+
+message NadaFunctionArg {
+  // Argument name
+  string name = 1;
+  // Argument type
+  nillion.nada.types.v1.NadaType type = 2;
+  // Source code info about this element.
+  uint64 source_ref_index = 3;
+}
+
+message NadaFunction {
+  // Function identifier
+  uint64 id = 1;
+  // Function arguments
+  repeated NadaFunctionArg args = 2;
+  // The name of the function
+  string name = 3;
+  // Table of operations for the function
+  map<uint64, nillion.nada.operations.v1.Operation>  operations = 4;
+  // Identifier of the operation (in the operations map) that represents
+  // the return of this function
+  uint64 return_operation_id = 5;
+  // Function return type
+  nillion.nada.types.v1.NadaType return_type = 6;
+  // NadaFunction source file information.
+  uint64 source_ref_index = 7;
+}
+
+message Party {
+  // Party index
+  uint64 id = 1;
+  // Name of the party
+  string name = 2;
+  // Source reference
+  uint64 source_ref_index = 3;
+}
+
+message Input {
+  // Input index
+  uint64 id = 1;
+  // Operation type
+  nillion.nada.types.v1.NadaType type = 2;
+  // Party that contains this input
+  string party = 3;
+  // Input name
+  string name = 4;
+  // The documentation.
+  string doc = 5;
+  // Source file info related with this operation.
+  uint64 source_ref_index = 6;
+}
+
+message Literal {
+  // Literal index
+  uint64 id = 1;
+  // Name
+  string name = 2;
+  // Value
+  string value = 3;
+  // Type
+  nillion.nada.types.v1.NadaType type = 4;
+}
+
+message SourceRef {
+  // Nada-lang file that contains the elements
+  string file = 1;
+  // Line number into the file that contains the element
+  uint32 lineno = 2;
+  // Element's offset into the file
+  uint32 offset = 3;
+  // Element's length into the file
+  uint32 length = 4;
+}
+
+message Output {
+  // Output name
+  string name = 1;
+  // Output operation ID
+  int64 operation_id = 2;
+  // Party contains this output
+  string party = 3;
+  // Output type
+  nillion.nada.types.v1.NadaType type = 4;
+  // Source file info related with this output.
+  int64 source_ref_index = 5;
+}
+
+// The Program MIR.
+// This is the container of all the MIR representation for a program.
+message ProgramMIR {
+  // List of the functions are used by the program
+  repeated NadaFunction functions = 1;
+  // Program parties
+  repeated Party parties = 2;
+  // Program inputs
+  repeated Input inputs = 3;
+  // Program literals
+  repeated Literal literals = 4;
+  // Program output
+  repeated Output outputs = 5;
+  // Table of operations
+  map<uint64, nillion.nada.operations.v1.Operation> operations = 6;
```

Comment:
Did you consider using a `repeated` instead of a map such that the index is the operation id? I imagine operation ids start at 0 so it would work from that angle. This would save at least 2 bytes per operation (see [reference](https://protobuf.dev/programming-guides/proto3/#backwards)).

Reply:
I tried to do it, but this requires more changes. Because operation IDs are shared between operations in ProgramMir and operations in Functions, I tried to create a new operation ID context for functions, but that broke other things, so I stopped there. I will consider changing it in a follow-up.